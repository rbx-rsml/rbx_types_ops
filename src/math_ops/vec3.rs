use rbx_types::{Variant, Vector3, Vector3int16};
use super::{BasicOperations, Operation, OperationFn};

fn operation_vector3_with_f32(
    left: &Vector3, right: f32, 
    operation_fn_f32: &OperationFn<f32>
) -> Vector3 {
    Vector3::new(
        operation_fn_f32(left.x, right),
        operation_fn_f32(left.y, right),
        operation_fn_f32(left.z, right),
    )
}

fn operation_vector3_with_vector3int16(
    left: &Vector3, right: &Vector3int16, 
    operation_fn_f32: &OperationFn<f32>
) -> Vector3 {
    Vector3::new(
        operation_fn_f32(left.x, right.x as f32),
        operation_fn_f32(left.y, right.y as f32),
        operation_fn_f32(left.z, right.z as f32),
    )
}

pub fn operation_vector3_with_vector3(
    left: &Vector3, right: &Vector3, 
    operation_fn_f32: &OperationFn<f32>
) -> Vector3 {
    Vector3::new(
        operation_fn_f32(left.x, right.x),
        operation_fn_f32(left.y, right.y),
        operation_fn_f32(left.z, right.z),
    )
}

impl Operation for Vector3 {
    fn operation(
        &self, with: &Variant,
        operation_fn_f32: OperationFn<f32>,
        _operation_fn_i32: OperationFn<i32>
    ) -> Option<Variant> {
        match with {
            Variant::Vector3(with) => Some(Variant::Vector3(operation_vector3_with_vector3(self, with, &operation_fn_f32))),
            Variant::Float32(with) => Some(Variant::Vector3(operation_vector3_with_f32(self, *with, &operation_fn_f32))),
            Variant::Vector3int16(with) => Some(Variant::Vector3(operation_vector3_with_vector3int16(self, with, &operation_fn_f32))),
            _ => None
        }
    }
}

impl BasicOperations for Vector3 {}