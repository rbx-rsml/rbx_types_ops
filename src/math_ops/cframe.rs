use rbx_types::{CFrame, Matrix3, Variant, Vector3};
use super::{BasicOperations, Operation, OperationFn};

use super::vec3::operation_vector3_with_vector3;

fn operation_cframe_with_cframe(
    left: &CFrame, right: &CFrame, 
    operation_fn_f32: &OperationFn<f32>
) -> CFrame {
    let left_orien = left.orientation;
    let right_orien = right.orientation;

    CFrame::new(
        operation_vector3_with_vector3(&left.position, &right.position, operation_fn_f32),
        Matrix3::new(
            operation_vector3_with_vector3(&left_orien.x, &right_orien.x, operation_fn_f32),
            operation_vector3_with_vector3(&left_orien.y, &right_orien.y, operation_fn_f32),
            operation_vector3_with_vector3(&left_orien.z, &right_orien.z, operation_fn_f32)
        )
    )
}

impl Operation for CFrame {
    fn operation(
        &self, with: &Variant,
        operation_fn_f32: OperationFn<f32>,
        _operation_fn_i32: OperationFn<i32>,
        _operation_fn_u8: OperationFn<u8>
    ) -> Option<Variant> {
        match with {
            Variant::CFrame(with) => Some(Variant::CFrame(operation_cframe_with_cframe(&self, &with, &operation_fn_f32))),
            _ => None
        }
    }
}

impl BasicOperations for CFrame {}


// cframe && cframe
#[test]
fn test_cframe_pow_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).pow(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}

#[test]
fn test_cframe_div_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).div(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}

#[test]
fn test_cframe_floor_div_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).floor_div(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}

#[test]
fn test_cframe_modulus_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).modulus(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}

#[test]
fn test_cframe_mult_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).mult(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}

#[test]
fn test_cframe_add_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).add(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}

#[test]
fn test_cframe_sub_cframe() {
    insta::assert_yaml_snapshot!(
        CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).sub(&CFrame::new(Vector3::new(5.0, 10.0, 15.0), Matrix3::new(
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0),
            Vector3::new(6.0, 12.0, 18.0)
        )).into())
    )
}