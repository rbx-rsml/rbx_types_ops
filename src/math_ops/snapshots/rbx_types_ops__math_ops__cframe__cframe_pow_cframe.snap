---
source: src/math_ops/cframe.rs
expression: "CFrame::new(Vector3::new(5.0, 10.0, 15.0),\nMatrix3::new(Vector3::new(6.0, 12.0, 18.0), Vector3::new(6.0, 12.0, 18.0),\nVector3::new(6.0, 12.0,\n18.0))).pow(&CFrame::new(Vector3::new(5.0, 10.0, 15.0),\nMatrix3::new(Vector3::new(6.0, 12.0, 18.0), Vector3::new(6.0, 12.0, 18.0),\nVector3::new(6.0, 12.0, 18.0))).into())"
---
CFrame:
  position:
    - 3125
    - 10000000000
    - 437893900000000000
  orientation:
    - - 46656
      - 8916100000000
      - 39346410000000000000000
    - - 46656
      - 8916100000000
      - 39346410000000000000000
    - - 46656
      - 8916100000000
      - 39346410000000000000000
